File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.cpp
class Solution 
{
public:

vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) 
    {
        vector<vector<int>> g(n);
        vector<vector<int>>ans(n);
        
        for(auto &i:edges)
        {
             g[i[0]].push_back(i[1]);
        }
        for(int i=0;i<n;i++){
            dfs(g,ans,i, i);
        }
        return ans;
    }
    
    void dfs(vector<vector<int>> &g,vector<vector<int>> &ans,int src, int anc)
    {
        for(auto &it : g[src])
        {
            if(ans[it].size()==0 || ans[it].back()!=anc)
            {
                ans[it].push_back(anc);
                dfs(g,ans,it,anc);
            }
        }
    }
};

File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.java
class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        
        HashMap<Integer, List<Integer>> ancestor = new HashMap<>();
        for(int i = 0; i < n;i++){
            ancestor.put(i,new ArrayList<>());
        }
        for(int[] arr : edges){
            List<Integer> list = ancestor.getOrDefault(arr[1], new ArrayList<>());
            list.add(arr[0]);
            ancestor.put(arr[1], list);
        }

        List<List<Integer>> ans = new ArrayList<>();
        for(int i = 0; i < n; i++){
            List<Integer> temp = ancestor.get(i);
            addAncestors(temp, ancestor,n);
            Collections.sort(temp);
            ans.add(temp);
        }
        return ans;
    }
    void addAncestors(List<Integer> list, HashMap<Integer, List<Integer>> ancestor, int n){
        if(list.size() == 0) return;

        boolean[] visited = new boolean[n]; 
        Queue<Integer> queue = new LinkedList<>();
        for(int i = 0; i < list.size(); i++){
            queue.offer(list.get(i));
            visited[list.get(i)] = true;
        }

        while(!queue.isEmpty()){
            int top = queue.poll();
            List<Integer> temp = ancestor.get(top);

            for(int node : temp){
                if(!visited[node]){
                    queue.offer(node);
                    visited[node] = true;
                    list.add(node);
                }
            }
        }
    }
}